go/proxyd/eth_rpc_types.go

```diff
@@ -0,0 +1,243 @@
+package proxyd
+
+import (
+	"reflect"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+func ptr[T any](n T) *T {
+	return &n
+}
+
+// TODO: also add in debug methods
+var argTypes = map[string][]reflect.Type{
```

Comment:
Going to need a test for each of these arg types and to guarantee that this contains all possible rpc requests that we plan on serving

Reply:
I wonder if we could reuse geth's rpc package and simply have receiver functions that have the expected types. It'll drastically simplify how methods are routed to functions. And then we could, use an ethclient to forward the RPC arguments to the backend. Doing this is out of the scope for this PR but something to consider.

Reply:
> I wonder if we could reuse geth's rpc package and simply have receiver functions that have the expected types

Do you mind elaborating a bit? These types are built dynamically in geth based on the functions declared in the `internal` package, so we can't import them unfortunately to do the same thing

> And then we could, use an ethclient to forward the RPC

One problem is that `ethclient` doesn't support all RPC methods, just the most common ones.

Reply:
I'm referring to [this](https://github.com/ethereum/go-ethereum/blob/6c3fea0fc9db561aafd9ec017980d0ac87516b87/node/rpcstack.go#L532) functionality
AFAICT, it should work for any function receiver, even when the function accepts an external type, like `RequestOptions` here.

Reply:
Are you suggesting to make a `[]rpc.API` where each element corresponds to each endpoint? Could you give an example of exactly what you mean?

Reply:
> Are you suggesting to make a []rpc.API where each element corresponds to each endpoint?

Pretty much, the same way the opnode implements `optimism` RPCs. Here's a [gist](https://gist.github.com/Inphi/45f57fc80f983fd959b0930853d93a7f) illustrating this.

Reply:
I like this idea quite a bit because it makes writing the handlers a lot easier. Would need to sort out the logic for which backend gets used inside of each handler, so really it would be a way to make managing the supported rpc methods better

Reply:
As I'm now implementing batch RPC I realize that it will be cumbersome to send JSON-RPC batches to upstream using the `rpc.API` interface.  This is because the `rpc.API` interface breaks up batched JSON-RPC requests into single request objects before the user-supplied API interface can handle them. I have some ideas on working around the limitation, but it'll be a bit too complicated to do this on this PR. Let's continue using the current approach for now.

I'll follow up with another PR that uses the `rpc.API` with upstream batch support.